File: rs/src/parsers/gradient.rs
use crate::parsers::basic_data_types::{parse_s_rgb8, parse_straight_s_rgba8};
use nom::le_u8 as parse_u8;
use nom::IResult;
use swf_tree as ast;

#[allow(unused_variables)]
pub fn parse_color_stop(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::ColorStop> {
  do_parse!(
    input,
    ratio: parse_u8
      >> color:
        switch!(value!(with_alpha),
          true => call!(parse_straight_s_rgba8) |
          false => map!(parse_s_rgb8, |c| ast::StraightSRgba8 {r: c.r, g: c.g, b: c.b, a: 255})
        )
      >> (ast::ColorStop {
        ratio: ratio,
        color: color,
      })
  )
}

#[allow(unused_variables)]
pub fn parse_gradient(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::Gradient> {
  do_parse!(
    input,
    flags: parse_u8
      >> spread_id: value!(flags >> 6)
      >> color_space_id: value!((flags & ((1 << 6) - 1)) >> 4)
      >> color_count: value!(flags & ((1 << 4) - 1))
      >> spread:
        switch!(value!(spread_id),
          0 => value!(ast::GradientSpread::Pad) |
          1 => value!(ast::GradientSpread::Reflect) |
          2 => value!(ast::GradientSpread::Repeat)
          // TODO: Default to error
        )
      >> color_space:
        switch!(value!(color_space_id),
          0 => value!(ast::ColorSpace::SRgb) |
          1 => value!(ast::ColorSpace::LinearRgb)
          // TODO: Default to error
        )
      >> colors: length_count!(value!(color_count), apply!(parse_color_stop, with_alpha))
      >> (ast::Gradient {
        spread: spread,
        color_space: color_space,
        colors: colors,
      })
  )
}

#[allow(unused_variables)]
pub fn parse_morph_color_stop(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::MorphColorStop> {
  do_parse!(
    input,
    start: apply!(parse_color_stop, with_alpha)
      >> end: apply!(parse_color_stop, with_alpha)
      >> (ast::MorphColorStop {
        ratio: start.ratio,
        color: start.color,
        morph_ratio: end.ratio,
        morph_color: end.color,
      })
  )
}

#[allow(unused_variables)]
pub fn parse_morph_gradient(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::MorphGradient> {
  do_parse!(
    input,
    flags: parse_u8
      >> spread_id: value!(flags >> 6)
      >> color_space_id: value!((flags & ((1 << 6) - 1)) >> 4)
      >> color_count: value!(flags & ((1 << 4) - 1))
      >> spread:
        switch!(value!(spread_id),
          0 => value!(ast::GradientSpread::Pad) |
          1 => value!(ast::GradientSpread::Reflect) |
          2 => value!(ast::GradientSpread::Repeat)
          // TODO: Default to error
        )
      >> color_space:
        switch!(value!(color_space_id),
          0 => value!(ast::ColorSpace::SRgb) |
          1 => value!(ast::ColorSpace::LinearRgb)
          // TODO: Default to error
        )
      >> colors: length_count!(value!(color_count), apply!(parse_morph_color_stop, with_alpha))
      >> (ast::MorphGradient {
        spread: spread,
        color_space: color_space,
        colors: colors,
      })
  )
}

File: ts/src/lib/parsers/gradient.ts
import { ReadableByteStream } from "@open-flash/stream";
import { Uint2, Uint4, Uint8 } from "semantic-types";
import { ColorSpace } from "swf-types/color-space";
import { ColorStop } from "swf-types/color-stop";
import { GradientSpread } from "swf-types/gradient-spread";
import { Gradient } from "swf-types/gradient";
import { MorphColorStop } from "swf-types/morph-color-stop";
import { MorphGradient } from "swf-types/morph-gradient";
import { StraightSRgba8 } from "swf-types/straight-s-rgba8";

import { parseSRgb8, parseStraightSRgba8 } from "./basic-data-types.js";

export function parseColorStop(byteStream: ReadableByteStream, withAlpha: boolean): ColorStop {
  const ratio: Uint8 = byteStream.readUint8();
  let color: StraightSRgba8;
  if (withAlpha) {
    color = parseStraightSRgba8(byteStream);
  } else {
    color = {...parseSRgb8(byteStream), a: 255};
  }
  return {ratio, color};
}

export function parseGradient(byteStream: ReadableByteStream, withAlpha: boolean): Gradient {
  const flags: Uint8 = byteStream.readUint8();
  // The spec says that spreadId and colorSpaceId should be ignored for shapeVersion < Shape4
  // and color count should be capped at 8. We're ignoring it for the moment.
  const spreadId: Uint2 = <Uint2> ((flags & ((1 << 8) - 1)) >>> 6);
  const colorSpaceId: Uint2 = <Uint2> ((flags & ((1 << 6) - 1)) >>> 4);
  const colorCount: Uint4 = <Uint4> (flags & ((1 << 4) - 1));
  let spread: GradientSpread;
  switch (spreadId) {
    case 0:
      spread = GradientSpread.Pad;
      break;
    case 1:
      spread = GradientSpread.Reflect;
      break;
    case 2:
      spread = GradientSpread.Repeat;
      break;
    default:
      throw new Error("Unexpected gradient spread");
  }
  let colorSpace: ColorSpace;
  switch (colorSpaceId) {
    case 0:
      colorSpace = ColorSpace.SRgb;
      break;
    case 1:
      colorSpace = ColorSpace.LinearRgb;
      break;
    default:
      throw new Error("Unexpected gradient spread");
  }
  const colors: ColorStop[] = [];
  for (let i: number = 0; i < colorCount; i++) {
    colors.push(parseColorStop(byteStream, withAlpha));
  }
  return {
    spread,
    colorSpace,
    colors,
  };
}

export function parseMorphColorStop(byteStream: ReadableByteStream, withAlpha: boolean): MorphColorStop {
  const {ratio, color} = parseColorStop(byteStream, withAlpha);
  const {ratio: morphRatio, color: morphColor} = parseColorStop(byteStream, withAlpha);
  return {ratio, color, morphRatio, morphColor};
}

export function parseMorphGradient(byteStream: ReadableByteStream, withAlpha: boolean): MorphGradient {
  const flags: Uint8 = byteStream.readUint8();
  const spreadId: Uint2 = <Uint2> ((flags & ((1 << 8) - 1)) >>> 6);
  const colorSpaceId: Uint2 = <Uint2> ((flags & ((1 << 6) - 1)) >>> 4);
  const colorCount: Uint4 = <Uint4> (flags & ((1 << 4) - 1));
  let spread: GradientSpread;
  switch (spreadId) {
    case 0:
      spread = GradientSpread.Pad;
      break;
    case 1:
      spread = GradientSpread.Reflect;
      break;
    case 2:
      spread = GradientSpread.Repeat;
      break;
    default:
      throw new Error("Unexpected gradient spread");
  }
  let colorSpace: ColorSpace;
  switch (colorSpaceId) {
    case 0:
      colorSpace = ColorSpace.SRgb;
      break;
    case 1:
      colorSpace = ColorSpace.LinearRgb;
      break;
    default:
      throw new Error("Unexpected gradient spread");
  }
  const colors: MorphColorStop[] = [];
  for (let i: number = 0; i < colorCount; i++) {
    colors.push(parseMorphColorStop(byteStream, withAlpha));
  }
  return {
    spread,
    colorSpace,
    colors,
  };
}
